### PR TITLE
Mass Payment proccessor

### PR DESCRIPTION
**Pull request by stevenjack**
### Problem

We want to be able to parse the IPN response from PayPal into a struct so we can easily access the data about the mass payment response.

### Solution

* Add `parser.go` that takes a request body, an item to be loaded and a processor for processing the form data and loading the item.

* Create `Values` struct to override default behaviour of `url.Values` and allow access to PayPal specific variables.

### Notes

- **Remember** to tag your merge commit after merging 👍

### CODE REVIEW

**Comment by stevenjack** [comment]
[[GPI 0.2.0] Adds processor and MassPayment struct](https://trello.com/c/Wg0sh1aR/936-gpi-0-2-0-adds-processor-and-masspayment-struct)

File: payload/mass_payment.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	MassPayment struct {
+		PaymentDate   time.Time         `form_field:"payment_date"`
+		PaymentStatus string            `form_field:"payment_status"`
+		Handling      []string          `form_field:"mc_handling"`
+		Items         []MassPaymentItem `form_field:"-"`
+	}
+
+	// MassPaymentItem contains details about a specific mass payment
+	// from a mass payment IPN response.
+	MassPaymentItem struct {
```
Comment (revett) [inline payload/mass_payment.go]: 🏄 Move to separate file.

File: payload/populate.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package payload
+
+import (
+	"reflect"
+	"strconv"
+)
+
+type (
+	// Populate takes a reflect.Type, reflect.Value, form data and optional key
+	// index and popultes the reflected struct with the form data.
```
Comment (revett) [inline payload/populate.go]: 🏄 `popultes `

File: payload/populate.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	// Processor interface for Populaste structs that process data.
+	Processor interface {
+		Process(itemType reflect.Type, itemValue reflect.Value, formData *Values, fieldIndex int, countField string)
+	}
+)
+
+// Process takes a structs type and value and recursively populates it based on the data
+// in formData. FieldIndex is used to provide the index for a field if that data is populated
+// from a field like 'qux_1 or qux1' and countField is the field used to work out how many fields
+// there are for fields like 'qux_1'.
```
Comment (revett) [inline payload/populate.go]: Improve this comment, as not clear.

File: payload/values.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	// into data accessible for a payload consuming
+	// the parser.
+	Values struct {
+		url.Values
+	}
+)
+
+const (
+	// suffixTypeNumericSuffix indicates that the indexing is
+	// a single numeric suffix for a key.
```
Comment (revett) [inline payload/values.go]: No need for comment.

File: payload/values.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	suffixTypeNumericSuffix = "suffixTypeNumericSuffix"
+
+	// suffixTypeUnderscoreNumericSuffix indicates that the indexing is
+	// an underscore numeric suffix (_x) for a key.
```
Comment (revett) [inline payload/values.go]: No need for comment.

**Review by revett — APPROVED** [review]
Mostly small comments about formatting etc or being clearer.

I still think we can be making our PRs smaller - especially on new services/packages. Helps a lot with reviewing 👍

**Comment by stevenjack** [comment]
@revett yea I'm trying to and an conscious this wasn't the easiest to parse. It's sometimes hard as well when it's a new repo to keep it small and succinct as things are in flux and change and before you know it you've got 10 files in the PR. I'll work on keeping them smaller 👍

**Pull request merged**